File: 5_dp/이준우/정수 삼각형.py
def solution(triangle):
    n = len(triangle)
    dp = [[0] * n for _ in range(n)]
    dp[0][0] = triangle[0][0]

    for i in range(1, n):
        for j in range(i + 1):
            dp[i][j] = triangle[i][j] + max(
                # j=i인 경우, 가장 오른쪽 -> [i - 1][i]는 존재하지 않는 값 따라서 i - 1이 최대임을 보장하기 위한 min
                dp[i - 1][min(j, i - 1)],
                # j=0인 경우, 가장 왼쪽이기 때문에 -1해도 0임을 보장하기 위한 max
                dp[i - 1][max(0, j - 1)],
            )

    return max(dp[n - 1])


"""
triangle[i][0 ~ i]

[[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]	
"""

File: 5_dp/김민주/정수삼각형.cpp
#include <string>
#include <vector>

using namespace std;

void dp(vector<vector<int>>& triangle, int a, int b){
    if (b == 0){
        triangle[a][b] += triangle[a-1][b]; 
    }else if (b == triangle[a].size()-1){
        triangle[a][b] += triangle[a-1][b-1]; 
    }else {
        triangle[a][b] += max(triangle[a-1][b-1], triangle[a-1][b]); 
    }
    
}

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int l = triangle.size();
    for (int i=1; i<l; i++){
        for (int j=0; j<triangle[i].size(); j++){
            dp(triangle, i, j);
        }
    }
    int l2 = triangle[l-1].size();

    for (int i=0; i<l2; i++){
        answer = max(triangle[l-1][i], answer);
    }
    return answer;
}